server/leafnode_proxy_test.go

```diff
@@ -0,0 +1,889 @@
+
+	// Create spoke server with proxy configuration including auth via config file
+	configContent := fmt.Sprintf(`
+		listen: "127.0.0.1:-1"
+		leafnodes {
+			reconnect_interval: "50ms"
+			remotes = [
+				{
+					url: "ws://127.0.0.1:%d"
+					proxy {
+						url: "%s"
+						username: "testuser"
+						password: "testpass"
+						timeout: 5s
+					}
+				}
+			]
+		}
+	`, hubOpts.Websocket.Port, proxy.url())
+
+	configFile := createConfFile(t, []byte(configContent))
+
+	spoke, _ := RunServerWithConfig(configFile)
+	defer spoke.Shutdown()
+
+	// Verify leafnode connections are established
+	checkLeafNodeConnected(t, spoke)
+	checkLeafNodeConnected(t, hub)
+}
+
+func TestLeafNodeHttpProxyTLSMismatchDetection(t *testing.T) {
+	// This test simulates the TLS mismatch scenario described in the feedback:
+	// - Leafnode configured with proxy but no TLS
+	// - Hub requires TLS
+	// - Connection should fail with appropriate error message
+
+	// Create hub server with TLS required using config file
+	hubConfig := createConfFile(t, []byte(`
+		listen: "127.0.0.1:-1"
+		leafnodes {
+			listen: "127.0.0.1:-1"
+			tls: {
+				cert_file: "../test/configs/certs/server-cert.pem"
+				key_file: "../test/configs/certs/server-key.pem"
+			}
+		}
+	`))
+
+	hub, hubOpts := RunServerWithConfig(hubConfig)
+	defer hub.Shutdown()
+
+	// Create HTTP proxy
+	proxy := createTestHTTPProxy(_EMPTY_, _EMPTY_)
+	proxy.start()
+	defer proxy.stop()
+
+	// Create spoke server with proxy but no TLS configuration (intentional mismatch)
+	spokeConfigContent := fmt.Sprintf(`
+		listen: "127.0.0.1:-1"
+		leafnodes {
+			reconnect_interval: "50ms"
+			remotes = [
+				{
+					url: "ws://127.0.0.1:%d"
+					proxy {
+						url: "%s"
+						timeout: 5s
+					}
+					# Intentionally no TLS configuration to create mismatch
+				}
+			]
+		}
+	`, hubOpts.LeafNode.Port, proxy.url())
+
+	spokeConfig := createConfFile(t, []byte(spokeConfigContent))
+	spoke, _ := RunServerWithConfig(spokeConfig)
+	defer spoke.Shutdown()
+
+	// Wait and verify that connection was NOT established due to TLS mismatch
+	// First attempt happens during RunServerWithConfig(), then retries every 50ms
+	time.Sleep(250 * time.Millisecond)
+
+	if spoke.NumLeafNodes() != 0 {
+		t.Errorf("Expected 0 leafnode connections due to TLS mismatch, got %d", spoke.NumLeafNodes())
+	}
+}
+
+func TestLeafNodeHttpProxyTunnelBasic(t *testing.T) {
+	// Create HTTP proxy with longer delay for robustness
+	proxy := createTestHTTPProxy(_EMPTY_, _EMPTY_)
+	proxy.setCloseDelay(200 * time.Millisecond)
+	proxy.start()
+	defer proxy.stop()
+
+	// Create a simple TCP server to connect to through proxy
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("Failed to create test server: %v", err)
+	}
+	defer listener.Close()
+
+	targetPort := listener.Addr().(*net.TCPAddr).Port
+	targetHost := fmt.Sprintf("127.0.0.1:%d", targetPort)
+
+	// Use WaitGroup to properly synchronize server response
+	var wg sync.WaitGroup
```

Comment:
I would like you to make the following changes because, using t.Fatal in a go routine would produce a race (if an error was really happening). The proposed approach solves not doing a t.Fatal/t.Error in a go routine, does the proper synchronization (with timeout) of the go routine (instead of having a `go func() { wg.Wait() close(done) }()`.

[fix.patch](https://github.com/user-attachments/files/22257851/fix.patch)